File: services/product-service/internal/config/rabbitmq.go
package config

import (
	"log"

	amqp "github.com/rabbitmq/amqp091-go"
)

const ProductEventsExchange = "product_events"

type RabbitMQ struct {
	Conn    *amqp.Connection
	Channel *amqp.Channel
}

func InitRabbitMQ(url string) *RabbitMQ {
	conn, err := amqp.Dial(url)
	if err != nil {
		log.Printf("warning: failed to connect to RabbitMQ: %v — events will not be published", err)
		return nil
	}

	ch, err := conn.Channel()
	if err != nil {
		log.Printf("warning: failed to open RabbitMQ channel: %v — events will not be published", err)
		conn.Close()
		return nil
	}

	if err := ch.ExchangeDeclare(
		ProductEventsExchange,
		"topic",
		true,
		false,
		false,
		false,
		nil,
	); err != nil {
		log.Printf("warning: failed to declare exchange: %v — events will not be published", err)
		ch.Close()
		conn.Close()
		return nil
	}

	log.Println("RabbitMQ connected and exchange declared")
	return &RabbitMQ{Conn: conn, Channel: ch}
}

func (r *RabbitMQ) Close() {
	if r == nil {
		return
	}
	if r.Channel != nil {
		r.Channel.Close()
	}
	if r.Conn != nil {
		r.Conn.Close()
	}
}

File: services/auth-service/src/config/rabbitmq.js
'use strict';

const amqplib = require('amqplib');

const EXCHANGE = process.env.RABBITMQ_USER_EVENTS_EXCHANGE || 'user_events';
const RABBITMQ_URL = process.env.RABBITMQ_URL || 'amqp://localhost';

let channel = null;

async function connectRabbitMQ() {
  try {
    const connection = await amqplib.connect(RABBITMQ_URL);

    connection.on('error', (err) => {
      console.error('RabbitMQ connection error:', err.message);
      channel = null;
    });

    connection.on('close', () => {
      console.warn('RabbitMQ connection closed.');
      channel = null;
    });

    channel = await connection.createChannel();
    await channel.assertExchange(EXCHANGE, 'topic', { durable: true });
    console.log(`RabbitMQ connected. Exchange "${EXCHANGE}" ready.`);
  } catch (err) {
    console.error('Failed to connect to RabbitMQ:', err.message);
    // Non-fatal: service continues without messaging
  }
}

function publishEvent(routingKey, payload) {
  if (!channel) {
    console.warn('RabbitMQ channel unavailable. Event not published:', routingKey);
    return;
  }
  try {
    const content = Buffer.from(JSON.stringify(payload));
    channel.publish(EXCHANGE, routingKey, content, {
      persistent: true,
      contentType: 'application/json',
      timestamp: Math.floor(Date.now() / 1000),
    });
    console.log(`Event published: ${routingKey}`);
  } catch (err) {
    console.error('Failed to publish event:', err.message);
  }
}

module.exports = { connectRabbitMQ, publishEvent };
